plugins/single_plugins/vswipe.cpp

```diff
@@ -76,6 +80,118 @@ class vswipe : public wf::plugin_interface_t
+        callback_win_up    = [=] (wf::activator_source_t, uint32_t) { return add_direction( 0, -1, get_top_view()); };
+        callback_win_down  = [=] (wf::activator_source_t, uint32_t) { return add_direction( 0,  1, get_top_view()); };
+
+        wf::option_wrapper_t<wf::activatorbinding_t> binding_left{"vswipe/binding_left"};
+        wf::option_wrapper_t<wf::activatorbinding_t> binding_right{"vswipe/binding_right"};
+        wf::option_wrapper_t<wf::activatorbinding_t> binding_up{"vswipe/binding_up"};
+        wf::option_wrapper_t<wf::activatorbinding_t> binding_down{"vswipe/binding_down"};
+
+        wf::option_wrapper_t<wf::activatorbinding_t> binding_win_left{"vswipe/binding_win_left"};
+        wf::option_wrapper_t<wf::activatorbinding_t> binding_win_right{"vswipe/binding_win_right"};
+        wf::option_wrapper_t<wf::activatorbinding_t> binding_win_up{"vswipe/binding_win_up"};
+        wf::option_wrapper_t<wf::activatorbinding_t> binding_win_down{"vswipe/binding_win_down"};
+
+        output->add_activator(binding_left,  &callback_left);
+        output->add_activator(binding_right, &callback_right);
+        output->add_activator(binding_up,    &callback_up);
+        output->add_activator(binding_down,  &callback_down);
+
+        output->add_activator(binding_win_left,  &callback_win_left);
+        output->add_activator(binding_win_right, &callback_win_right);
+        output->add_activator(binding_win_up,    &callback_win_up);
+        output->add_activator(binding_win_down,  &callback_win_down);
+    }
+
+    wayfire_view get_top_view()
+    {
+        auto ws = output->workspace->get_current_workspace();
+        auto views = output->workspace->get_views_on_workspace(ws,
+            wf::LAYER_WORKSPACE | wf::LAYER_FULLSCREEN, true);
+
+        return views.empty() ? nullptr : views[0];
+    }
+
+    template<class wlr_event> using event = wf::input_event_signal<wlr_event>;
+
+    bool add_direction(int x, int y, wayfire_view view = nullptr)
+    {
+        if (!x && !y)
+            return false;
+
+        if (state.animating)
+            return false;
+
+        if (view && view->role != wf::VIEW_ROLE_TOPLEVEL)
+            view = nullptr;
+
+        event<wlr_event_pointer_swipe_begin> swipe_begin;
+        struct wlr_event_pointer_swipe_begin swipe_begin_ev = {
+                .device = nullptr,
+                .time_msec = 0,
+                .fingers = (uint32_t) fingers,
+        };
+        swipe_begin.event = &swipe_begin_ev;
+
+        event<wlr_event_pointer_swipe_update> swipe_update;
+        struct wlr_event_pointer_swipe_update swipe_update_ev = {
+                .device = nullptr,
+                .time_msec = 0,
+                .fingers = 0,
+                .dx = x * -100,
+                .dy = y * -100,
+        };
+        swipe_update.event = &swipe_update_ev;
+
+        on_swipe_begin(&swipe_begin);
```

Comment:
Would be nice if you could extract the functionality of `swipe_begin`, etc. to a separate function which is called both here and at the other place.

Simulating events is usually a hack which in this case can be avoided quite easily.